### ISSUE TITLE
Dataloaders for PyTorch

### ISSUE THREAD

**Issue by hhsecond** [tensorwerk/hangar-py#13]
@rlizzo I was thinking an API like `hangar.dataloaders.pytorch` (let me know if you have another structuring in your mind). Basically, the idea is to load data in batches synchronously or asynchronously and enable the features of PyTorch [DataLoader](https://pytorch.org/docs/stable/data.html#torch.utils.data.DataLoader). My plan is to have a Dataset class and a DataLoader class which is essentially the way PyTorch's data loader work.

**Comment by rlizzo** [comment]
So before we go ahead with this we should have a batched get method for the regular dataset method. If we get that right we can keep most of the functionality in the hangar base while making it really easy to plugin to the pytorch and tensorflow data loader framework

**Comment by hhsecond** [comment]
@rlizzo I made the skeleton of the dataloader base class #73 . Adding features such as batching, shuffling to the base dataloader is not required as of now since both PyTorch and Tensorflow does that quite well. They just need an object from where they could get one sample at a time. In my view, the base class can be inherited and make advanced dataloaders inside hangar itself when it's required in the future.

**Comment by rlizzo** [comment]
Hey @hhsecond it's going to be time to revisit this soon. I have a Dev branch where Read checkouts are thread and process safe (which includes a batched loader as well). I'm going to propose it in a PR in the next day or so, but we should coordinate on how this will plug in.

**Comment by hhsecond** [comment]
Sure thing. I'll have a look into that PR before we could get into a call or something

**Comment by hhsecond** [comment]
@rlizzo Do you have a preferred directory structure for including `dataloaders`? Also, I am thinking of the API as `hangar.dataloaders.TorchLoader`

**Comment by rlizzo** [comment]
I don't have a preferred structure right now. Let's take a look a simple implementation and we can restructure as needed from there. 

I also like the `hangar.dataloaders.foo` for now. Since the use facing functions are very limited, we can probably just hoist the `TorchLoader` / `TensorflowLoader` functions up to the top level (`from hangar import TorchLoarder`) but that can happen with `__init__.py` magic later.

**Comment by hhsecond** [comment]
If you want use init file for importing and stuff, we'll have to keep the data loader folder inside hangar folder (importing a level above the package is not recommended in python although we can work around it)

**Comment by rlizzo** [comment]
Good point. We could get fancy with the packages spec in `setup.py`, but it's probably not worth the hastle.

Why don't you go ahead and put it in the `src/hangar` directory for now, and we can figure out where it lives once we know what everything looks like?

**Comment by rlizzo** [comment]
Closed in #91
